Multi-Channel and Edge32/128BitConfigs Fail Diplomatic Elaboration

<!-- THIS REPOSITORY'S ISSUES ARE RESERVED FOR FEATURE REQUESTS AND BUG REPORTS. -->
<!-- For support questions related to Rocket Chip, please use [Stack Overflow [rocket-chip]](https://stackoverflow.com/questions/tagged/rocket-chip). -->
<!-- For support questions related to Chisel, please use [Stack Overflow [chisel]](https://stackoverflow.com/questions/tagged/chisel). -->
<!-- Please select the item best describing the issue in each category and delete the other items. -->

<!-- choose all that apply -->
**Type of issue**: bug report

<!-- choose one -->
**Impact**: no functional change

<!-- choose one -->
**Development Phase**: request

**Other information**
 <!-- include detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. Stack Overflow, gitter, etc -->

Some of the following configurations included in `src/main/scala/system/Configs.scala` fail during diplomatic elaboration with the associated requirements failing:
- `DualChannelConfig`
  - `dut.memAXI4Node (A sink node with parent dut) has only 1 inputs connected out of 2`
- `DualChannelDualBankConfig`
  - `dut.memAXI4Node (A sink node with parent dut) has only 1 inputs connected out of 2`
- `EightChannelConfig`
  - `dut.memAXI4Node (A sink node with parent dut) has only 1 inputs connected out of 8`

The common denominator here seems to be any configuration that mixes in `WithNMemoryChannels(n) if (n > 1)`.

Additionally, `Edge128BitConfig` and `Edge32BitConfig` fail, but those have different failing requirements:

- `Edge32BitConfig`
  - `xbar.intnode (A nexus node with parent xbar inside dut) has 1 required outputs and no possible inputs`
- `Edge128BitConfig`
  - `xbar.intnode (A nexus node with parent xbar inside dut) has 1 required outputs and no possible inputs`

Example output:
```
> make verilog -j4 CONFIG=DualChannelConfig
mkdir -p /home/se/repos/github.com/freechipsproject/rocket-chip/vsim/generated-src/
cd /home/se/repos/github.com/freechipsproject/rocket-chip && java -Xmx2G -Xss8M -XX:MaxPermSize=256M -jar /home/se/repos/github.com/freechipsproject/rocket-chip/sbt-launch.jar ++2.12.4 "runMain freechips.rocketchip.system.Generator /home/se/repos/github.com/freechipsproject/rocket-chip/vsim/generated-src freechips.rocketchip.system TestHarness freechips.rocketchip.system DualChannelConfig"
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=256M; support was removed in 8.0
[info] Loading settings for project global-plugins from plugins.sbt ...
[info] Loading global plugins from /home/se/.sbt/1.0/plugins
[info] Loading settings for project rocket-chip-build from plugins.sbt ...
[info] Loading project definition from /home/se/repos/github.com/freechipsproject/rocket-chip/project
[info] Loading settings for project rocketchip from build.sbt ...
[info] Loading settings for project hardfloat from build.sbt ...
[info] Loading settings for project chisel from build.sbt ...
Using addons: 
[info] Set current project to rocketchip (in build file:/home/se/repos/github.com/freechipsproject/rocket-chip/)
[info] Setting Scala version to 2.12.4 on 6 projects.
[info] Reapplying settings...
Using addons: 
[info] Set current project to rocketchip (in build file:/home/se/repos/github.com/freechipsproject/rocket-chip/)
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list
[info] Running freechips.rocketchip.system.Generator /home/se/repos/github.com/freechipsproject/rocket-chip/vsim/generated-src freechips.rocketchip.system TestHarness freechips.rocketchip.system DualChannelConfig
[info] [0.007] Elaborating design...
[error] (run-main-0) java.lang.reflect.InvocationTargetException
[error] java.lang.reflect.InvocationTargetException
[error] 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[error] 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[error] 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
[error] 	at freechips.rocketchip.util.HasGeneratorUtilities.$anonfun$elaborate$1(GeneratorUtils.scala:54)
[error] 	at chisel3.core.Module$.do_apply(Module.scala:51)
[error] 	at chisel3.Driver$.$anonfun$elaborate$1(Driver.scala:94)
[error] 	at chisel3.internal.Builder$.$anonfun$build$2(Builder.scala:317)
[error] 	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
[error] 	at chisel3.internal.Builder$.$anonfun$build$1(Builder.scala:315)
[error] 	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
[error] 	at chisel3.internal.Builder$.build(Builder.scala:315)
[error] 	at chisel3.Driver$.elaborate(Driver.scala:94)
[error] 	at freechips.rocketchip.util.HasGeneratorUtilities.elaborate(GeneratorUtils.scala:59)
[error] Caused by: java.lang.IllegalArgumentException: requirement failed: dut.memAXI4Node (A sink node with parent dut) has only 1 inputs connected out of 2
[error] 	at scala.Predef$.require(Predef.scala:277)
[error] 	at freechips.rocketchip.diplomacy.SinkNode.resolveStar(Nodes.scala:522)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.liftedTree1$1(Nodes.scala:270)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.x$26$lzycompute(Nodes.scala:255)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.x$26(Nodes.scala:254)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.iPortMapping$lzycompute(Nodes.scala:254)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.iPortMapping(Nodes.scala:254)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.iPortMapping(Nodes.scala:220)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.$anonfun$oPorts$1(Nodes.scala:290)
[error] 	at scala.collection.immutable.List.flatMap(List.scala:335)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.oPorts$lzycompute(Nodes.scala:289)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.oPorts(Nodes.scala:289)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.uoParams$lzycompute(Nodes.scala:313)
[error] 	at freechips.rocketchip.diplomacy.MixedNode.uoParams(Nodes.scala:313)
[error] Nonzero exit code: 1
[error] (Compile / runMain) Nonzero exit code: 1
[error] Total time: 5 s, completed Jan 15, 2019 1:45:58 PM
/home/se/repos/github.com/freechipsproject/rocket-chip/vsim/Makefrag-verilog:12: recipe for target '/home/se/repos/github.com/freechipsproject/rocket-chip/vsim/generated-src/freechips.rocketchip.system.DualChannelConfig.fir' failed
make: *** [/home/se/repos/github.com/freechipsproject/rocket-chip/vsim/generated-src/freechips.rocketchip.system.DualChannelConfig.fir] Error 1
```

**If the current behavior is a bug, please provide the steps to reproduce the problem:**

```bash
cd vsim
make CONFIG=DualChannelConfig
```

**What is the current behavior?**

Diplomatic requirement fails as only the first channel is getting connected.

**What is the expected behavior?**

This shouldn't error.

**Please tell us about your environment:**
<!--    (examples)

- git hash: ad61beed8a4c7593bc91f0c8ea51e927616b50f8
-->
**What is the use case for changing the behavior?**

Shipped default configurations should probably work. This is likely similar to the `RoccExampleConfig` situation where all top-level system configurations should be tested that they can get all the way to Verilog.

--- a/README.md
+++ b/README.md
@@ -401,18 +401,6 @@ BaseConfig. You can change those numbers to get a Rocket core with different
 cache parameters. For example, by changing L1I, NWays to 4, you will get
 a 32KB 4-way set-associative L1 instruction cache rather than a 16KB 2-way
 set-associative L1 instruction cache.
-
-Further down, you will be able to see two FPGA configurations:
-DefaultFPGAConfig and DefaultFPGASmallConfig. DefaultFPGAConfig inherits from
-BaseConfig, but overrides the low-performance memory port (i.e., backup
-memory port) to be turned off. This is because the high-performance memory
-port is directly connected to the high-performance AXI interface on the ZYNQ
-FPGA. DefaultFPGASmallConfig inherits from DefaultFPGAConfig, but changes the
-cache sizes, disables the FPU, turns off the fast early-out multiplier and
-divider, and reduces the number of TLB entries (all defined in SmallConfig).
-This small configuration is used for the Zybo FPGA board, which has the
-smallest ZYNQ part.
-
 Towards the end, you can also find that DefaultSmallConfig inherits all
 parameters from BaseConfig but overrides the same parameters of
 WithNSmallCores.

--- a/regression/Makefile
+++ b/regression/Makefile
@@ -82,9 +82,25 @@ CONFIGS += $(CONFIGS_32)
 CONFIGS += $(CONFIGS_64)
 endif
 
+# Miscellaneous configs cover any remaining configurations not tested
+# above, but are included in the freechips.rocketchip.system package.
+# These are here to prevent regressions at the compilation level and
+# expected to be built, but not executed.
 ifeq ($(SUITE), Miscellaneous)
 PROJECT=freechips.rocketchip.system
-CONFIGS=RoccExampleConfig
+CONFIGS=\
+	DefaultSmallConfig \
+	DualBankConfig \
+	DualChannelConfig \
+	DualChannelDualBankConfig \
+	RoccExampleConfig \
+	Edge128BitConfig \
+	Edge32BitConfig \
+	QuadChannelBenchmarkConfig \
+	EightChannelConfig \
+	DualCoreConfig \
+	MemPortOnlyConfig \
+	MMIOPortOnlyConfig
 endif
 
 # These are the named regression targets.  While it's expected you run them in

--- a/src/main/scala/subsystem/MemoryBus.scala
+++ b/src/main/scala/subsystem/MemoryBus.scala
@@ -55,7 +55,7 @@ class MemoryBus(params: MemoryBusParams)(implicit p: Parameters)
 
   private val xbar = LazyModule(new TLXbar).suggestName(busName + "_xbar")
   def inwardNode: TLInwardNode =
-    if (params.replicatorMask == 0) xbar.node else { xbar.node :*= RegionReplicator(params.replicatorMask) }
+    if (params.replicatorMask == 0) xbar.node else { xbar.node :=* RegionReplicator(params.replicatorMask) }
   def outwardNode: TLOutwardNode = ProbePicker() :*= xbar.node
   def busView: TLEdge = xbar.node.edges.in.head
   attachBuiltInDevices(params)

--- a/src/main/scala/subsystem/MemoryBus.scala
+++ b/src/main/scala/subsystem/MemoryBus.scala
@@ -64,6 +64,6 @@ class MemoryBus(params: MemoryBusParams)(implicit p: Parameters)
       (name: Option[String] = None, buffer: BufferParams = BufferParams.none)
       (gen: => NodeHandle[ TLClientPortParameters,TLManagerPortParameters,TLEdgeIn,TLBundle, D,U,E,B] =
         TLNameNode(name)): OutwardNodeHandle[D,U,E,B] = {
-    to("memory_controller" named name) { gen := TLWidthWidget(params.beatBytes) := TLBuffer(buffer) := outwardNode }
+    to("memory_controller" named name) { gen :*= TLWidthWidget(params.beatBytes) :*= TLBuffer(buffer) :*= outwardNode }
   }
 }

--- a/src/main/scala/subsystem/Ports.scala
+++ b/src/main/scala/subsystem/Ports.scala
@@ -52,8 +52,10 @@ trait CanHaveMasterAXI4MemPort { this: BaseSubsystem =>
         beatBytes = memPortParams.beatBytes)
     })
 
-    memAXI4Node := mbus.toDRAMController(Some(portName)) {
-      AXI4UserYanker() := AXI4IdIndexer(memPortParams.idBits) := TLToAXI4()
+    (0 until nMemoryChannels).map { i => 
+      memAXI4Node := mbus.toDRAMController(Some(portName)) {
+        AXI4UserYanker() := AXI4IdIndexer(memPortParams.idBits) := TLToAXI4()
+      }
     }
 
     memAXI4Node

--- a/src/main/scala/subsystem/Ports.scala
+++ b/src/main/scala/subsystem/Ports.scala
@@ -52,10 +52,8 @@ trait CanHaveMasterAXI4MemPort { this: BaseSubsystem =>
         beatBytes = memPortParams.beatBytes)
     })
 
-    (0 until nMemoryChannels).map { i => 
-      memAXI4Node := mbus.toDRAMController(Some(portName)) {
-        AXI4UserYanker() := AXI4IdIndexer(memPortParams.idBits) := TLToAXI4()
-      }
+    memAXI4Node :*= mbus.toDRAMController(Some(portName)) {
+      AXI4UserYanker() :*= AXI4IdIndexer(memPortParams.idBits) :*= TLToAXI4()
     }
 
     memAXI4Node

--- a/src/main/scala/system/Configs.scala
+++ b/src/main/scala/system/Configs.scala
@@ -80,6 +80,4 @@ class MMIOPortOnlyConfig extends Config(
 )
 
 class BaseFPGAConfig extends Config(new BaseConfig)
-
 class DefaultFPGAConfig extends Config(new WithNSmallCores(1) ++ new BaseFPGAConfig)
-class DefaultFPGASmallConfig extends Config(new DefaultFPGAConfig)

--- a/src/main/scala/system/Configs.scala
+++ b/src/main/scala/system/Configs.scala
@@ -31,13 +31,13 @@ class DefaultSmallConfig extends Config(new WithNSmallCores(1) ++ new BaseConfig
 class DefaultRV32Config extends Config(new WithRV32 ++ new DefaultConfig)
 
 class DualBankConfig extends Config(
-  new WithNBanks(2) ++ new BaseConfig)
+  new WithNBanks(2) ++ new DefaultConfig)
 
 class DualChannelConfig extends Config(new WithNMemoryChannels(2) ++ new DefaultConfig)
 
 class DualChannelDualBankConfig extends Config(
   new WithNMemoryChannels(2) ++
-  new WithNBanks(4) ++ new BaseConfig)
+  new WithNBanks(4) ++ new DefaultConfig)
 
 class RoccExampleConfig extends Config(new WithRoccExample ++ new DefaultConfig)
 
@@ -51,7 +51,7 @@ class DualChannelBenchmarkConfig extends Config(new WithNMemoryChannels(2) ++ ne
 class QuadChannelBenchmarkConfig extends Config(new WithNMemoryChannels(4) ++ new SingleChannelBenchmarkConfig)
 class OctoChannelBenchmarkConfig extends Config(new WithNMemoryChannels(8) ++ new SingleChannelBenchmarkConfig)
 
-class EightChannelConfig extends Config(new WithNMemoryChannels(8) ++ new BaseConfig)
+class EightChannelConfig extends Config(new WithNMemoryChannels(8) ++ new DefaultConfig)
 
 class DualCoreConfig extends Config(
   new WithNBigCores(2) ++ new BaseConfig)

--- a/src/main/scala/system/Configs.scala
+++ b/src/main/scala/system/Configs.scala
@@ -33,7 +33,7 @@ class DefaultRV32Config extends Config(new WithRV32 ++ new DefaultConfig)
 class DualBankConfig extends Config(
   new WithNBanks(2) ++ new BaseConfig)
 
-class DualChannelConfig extends Config(new WithNMemoryChannels(2) ++ new BaseConfig)
+class DualChannelConfig extends Config(new WithNMemoryChannels(2) ++ new DefaultConfig)
 
 class DualChannelDualBankConfig extends Config(
   new WithNMemoryChannels(2) ++
@@ -42,9 +42,9 @@ class DualChannelDualBankConfig extends Config(
 class RoccExampleConfig extends Config(new WithRoccExample ++ new DefaultConfig)
 
 class Edge128BitConfig extends Config(
-  new WithEdgeDataBits(128) ++ new BaseConfig)
+  new WithEdgeDataBits(128) ++ new DefaultConfig)
 class Edge32BitConfig extends Config(
-  new WithEdgeDataBits(32) ++ new BaseConfig)
+  new WithEdgeDataBits(32) ++ new DefaultConfig)
 
 class SingleChannelBenchmarkConfig extends Config(new DefaultConfig)
 class DualChannelBenchmarkConfig extends Config(new WithNMemoryChannels(2) ++ new SingleChannelBenchmarkConfig)

--- a/src/main/scala/system/Configs.scala
+++ b/src/main/scala/system/Configs.scala
@@ -24,7 +24,7 @@ class BaseConfig extends Config(
 
 class DefaultConfig extends Config(new WithNBigCores(1) ++ new BaseConfig)
 
-class DefaultBufferlessConfig extends Config(new WithBufferlessBroadcastHub ++ DefaultConfig)
+class DefaultBufferlessConfig extends Config(new WithBufferlessBroadcastHub ++ new DefaultConfig)
 class DefaultSmallConfig extends Config(new WithNSmallCores(1) ++ new BaseConfig)
 class DefaultRV32Config extends Config(new WithRV32 ++ new DefaultConfig)
 

--- a/src/main/scala/system/Configs.scala
+++ b/src/main/scala/system/Configs.scala
@@ -24,16 +24,14 @@ class BaseConfig extends Config(
 
 class DefaultConfig extends Config(new WithNBigCores(1) ++ new BaseConfig)
 
-class DefaultBufferlessConfig extends Config(
-  new WithBufferlessBroadcastHub ++ new WithNBigCores(1) ++ new BaseConfig)
-
+class DefaultBufferlessConfig extends Config(new WithBufferlessBroadcastHub ++ DefaultConfig)
 class DefaultSmallConfig extends Config(new WithNSmallCores(1) ++ new BaseConfig)
 class DefaultRV32Config extends Config(new WithRV32 ++ new DefaultConfig)
 
-class DualBankConfig extends Config(
-  new WithNBanks(2) ++ new DefaultConfig)
-
+class DualBankConfig extends Config(new WithNBanks(2) ++ new DefaultConfig)
+class DualCoreConfig extends Config( new WithNBigCores(2) ++ new BaseConfig)
 class DualChannelConfig extends Config(new WithNMemoryChannels(2) ++ new DefaultConfig)
+class EightChannelConfig extends Config(new WithNMemoryChannels(8) ++ new DefaultConfig)
 
 class DualChannelDualBankConfig extends Config(
   new WithNMemoryChannels(2) ++
@@ -50,11 +48,6 @@ class SingleChannelBenchmarkConfig extends Config(new DefaultConfig)
 class DualChannelBenchmarkConfig extends Config(new WithNMemoryChannels(2) ++ new SingleChannelBenchmarkConfig)
 class QuadChannelBenchmarkConfig extends Config(new WithNMemoryChannels(4) ++ new SingleChannelBenchmarkConfig)
 class OctoChannelBenchmarkConfig extends Config(new WithNMemoryChannels(8) ++ new SingleChannelBenchmarkConfig)
-
-class EightChannelConfig extends Config(new WithNMemoryChannels(8) ++ new DefaultConfig)
-
-class DualCoreConfig extends Config(
-  new WithNBigCores(2) ++ new BaseConfig)
 
 class TinyConfig extends Config(
   new WithNoMemPort ++

